libindy_vdr/src/pool/cache/mod.rs

```diff
@@ -0,0 +1,65 @@
+use async_lock::RwLock;
+use async_trait::async_trait;
+use std::{fmt::Display, sync::Arc};
+
+pub mod storage;
+pub mod strategy;
+
+#[async_trait]
+pub trait CacheStrategy<K, V>: Send + Sync + 'static {
+    async fn get(&self, key: &K) -> Option<V>;
+
+    async fn remove(&mut self, key: &K) -> Option<V>;
+
+    async fn insert(&mut self, key: K, value: V, custom_exp_offset: Option<u128>) -> Option<V>;
+}
+
+pub struct Cache<K, V> {
+    storage: Arc<RwLock<dyn CacheStrategy<String, V>>>,
+    key_prefix: Option<K>,
+}
+
+impl<K: Display + 'static, V: 'static> Cache<K, V> {
+    fn full_key(&self, key: &K) -> String {
+        match &self.key_prefix {
+            Some(prefix) => format!("{}{}", prefix, key),
+            None => key.to_string(),
+        }
+    }
+
+    pub fn new(storage: impl CacheStrategy<String, V>, key_prefix: Option<K>) -> Self {
+        Self {
+            storage: Arc::new(RwLock::new(storage)),
+            key_prefix,
+        }
+    }
+
+    pub async fn get(&self, key: &K) -> Option<V> {
+        let full_key = self.full_key(key);
+        self.storage.read().await.get(&full_key).await
+    }
+
+    pub async fn remove(&self, key: &K) -> Option<V> {
+        let full_key = self.full_key(key);
+        self.storage.write().await.remove(&full_key).await
+    }
+
+    pub async fn insert(&self, key: K, value: V, custom_exp_offset: Option<u128>) -> Option<V> {
```

Comment:
Have never used it myself but do we know the support for `u128`? Is this something that LLVM fixes for us?

Reply:
In this thread here: https://stackoverflow.com/questions/57340308/how-does-rusts-128-bit-integer-i128-work-on-a-64-bit-system 
it looks like u128s are compiled down to LLVM and handled by multiple registers